geth/swarm.go

```diff
@@ -0,0 +1,186 @@
+package geth
+
+import (
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/contracts/chequebook"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/swarm"
+	bzzapi "github.com/ethereum/go-ethereum/swarm/api"
+	"path/filepath"
+	"sync"
+)
+
+var (
+	ErrInvalidSwarmService           = errors.New("swarm service is unavailable")
+	ErrSwarmIdentityInjectionFailure = errors.New("failed to inject identity into Swarm")
+)
+
+type SwarmService struct {
+	swarm  *swarm.Swarm
+	keys   map[string]*ecdsa.PrivateKey
+	keysMu sync.RWMutex // Mutex to sync identity keys
+}
+
+func NewSwarmService() *SwarmService {
+	service := &SwarmService{
+		keys: make(map[string]*ecdsa.PrivateKey),
+	}
+
+	return service
+}
+
+// activateEthService configures and registers the eth.Ethereum service with a given node.
+func (self *SwarmService) StartSwarm(prvkey *ecdsa.PrivateKey) error {
+	if self.swarm != nil {
+		err := self.Stop()
+		if err != nil {
+			return err
+		}
+
+	}
+
+	nm := NodeManagerInstance()
+	stack := nm.node.geth
+
+	chbookaddr := common.Address{}
+	bzzdir := filepath.Join(stack.InstanceDir(), "swarm")
+
+	networkId := uint64(1)
+	if UseTestnet {
+		networkId = 3
+	}
+
+	bzzconfig, err := bzzapi.NewConfig(bzzdir, chbookaddr, prvkey, networkId)
+	if err != nil {
+		return fmt.Errorf("%v: %v", err, err)
+	}
+
+	var client *ethclient.Client
+	client, err = ethclient.Dial(fmt.Sprintf("http://%s", stack.HTTPEndpoint()))
+	if err != nil {
+		return fmt.Errorf("Can't connect: %v", err)
+	}
+
+	self.swarm, err = swarm.NewSwarm(nil, client, bzzconfig, false, true, "")
+	if err != nil {
+		return err
+	}
+
+	return self.Start(nm.node.geth.Server())
+}
+
+func (self *SwarmService) Start(net *p2p.Server) error {
+	if self.swarm != nil && net != nil {
+		return self.swarm.Start(net)
+	}
+
+	return nil
+}
+
+func (self *SwarmService) Stop() error {
+	if self.swarm != nil {
+		err := self.swarm.Stop()
+		if err != nil {
+			return err
+		}
+
+		self.swarm = nil
+	}
+
+	return nil
+}
+
+func (self *SwarmService) Protocols() []p2p.Protocol {
```

Comment:
**Important:**
- The problem with "let's wrap everything into a proxy" is that we will need to keep proxy in sync with underlying implementation.
- Ideally, in future swarm will allow changing running private key on fly, which will simplify our code even further. And we even can make this happen, as nothing prevents us from modifying our fork of `go-ethereum`
- Until that, let's try to keep our swarm management code to the minimally possible, w/o reindexing public/private protocol swarm exposes via our proxies.

Basically, all we need is to start/stop service by feeding into it some parameters (like private key).